### PR TITLE
feat: Oura Ring app

### PR DESCRIPTION
**Pull request by leog**
## What does this PR do?

Implementing Oura Ring app. Here is the description:

> Rely on your Oura Ring Readiness score to allow bookings on the days you have data loaded from your ring. A score equal to or greater than 70, means you've recovered well enough so bookings are not blocked, below that, you need to pay attention as you're not fully recovered, so bookings will be blocked. No data means bookings are not blocked.

<kbd></kbd>

## Type of change

- New feature (non-breaking change which adds functionality)

## How should this be tested?

Install the app following the instructions to create a client id and secret and expect availability to be impacted by readiness below 70.

## Mandatory Tasks

- [x] Make sure you have self-reviewed the code. A decent size PR without self-review might be rejected.

### CODE REVIEW

File: .env.appStore.example [hunk 1]
```diff
@@ -53,6 +54,12 @@ HUBSPOT_CLIENT_SECRET=""
 MS_GRAPH_CLIENT_ID=
 MS_GRAPH_CLIENT_SECRET=
 
+#   - OURA RING
+# Used for the Oura Ring integration
+# @see https://github.com/calcom/cal.com/#Obtaining-Oura-Ring-Client-ID-and-Secret
+OURA_RING_CLIENT_ID=
+OURA_RING_CLIENT_SECRET=
+
```
Comment (leog) [inline .env.appStore.example:62]: Adding environment variables for Oura Ring OAuth2 connection

File: README.md [hunk 2]
```diff
@@ -475,6 +475,14 @@ following
...
 10. You're good to go. Now you can see any booking in Cal.com created as a meeting in HubSpot for your contacts.
 
+## Obtaining Oura Ring Client ID and Secret
+1. Go to [Oura Ring Developer portal](https://cloud.ouraring.com/user/sign-in?next=%2Foauth%2Fapplications)
+2. Click button "New application"
+3. Fill in the information for your new Cal.com Oura Ring application 
+7. Set the Redirect URL to `<Cal.com URL>/api/integrations/ouraring/callback` replacing Cal.com URL with the URI at which your application runs.
+9. Click the "Save" button at the bottom.
+6. Now copy the Client ID and Client Secret to your `.env` file into the `OURA_RING_CLIENT_ID` and `OURA_RING_CLIENT_SECRET` fields.
+
```
Comment (leog) [inline README.md:485]: Instructions to get client id and secret from Oura Ring platform

File: packages/app-store/apps.metadata.generated.ts [hunk 3]
```diff
@@ -96,6 +97,7 @@ export const appStoreMetadata = {
   n8n: n8n_config_json,
   office365calendar: office365calendar__metadata_ts,
   office365video: office365video_config_json,
+  ouraring: ouraring_config_json,
   ping: ping_config_json,
   pipedream: pipedream_config_json,
   plausible: plausible_config_json,
```
Comment (leog) [inline packages/app-store/apps.metadata.generated.ts:103]: Generated

File: packages/app-store/apps.server.generated.ts [hunk 4]
```diff
@@ -31,6 +31,7 @@ export const apiHandlers = {
   n8n: import("./n8n/api"),
   office365calendar: import("./office365calendar/api"),
   office365video: import("./office365video/api"),
+  ouraring: import("./ouraring/api"),
   ping: import("./ping/api"),
   pipedream: import("./pipedream/api"),
```
Comment (leog) [inline packages/app-store/apps.server.generated.ts:36]: Generated

File: packages/app-store/index.ts [hunk 5]
```diff
@@ -13,6 +13,7 @@ const appStore = {
   larkcalendar: () => import("./larkcalendar"),
   office365calendar: () => import("./office365calendar"),
   office365video: () => import("./office365video"),
+  ouraring: () => import("./ouraring"),
   plausible: () => import("./plausible"),
   salesforce: () => import("./salesforce"),
```
Comment (leog) [inline packages/app-store/index.ts:18]: Generated

File: packages/app-store/ouraring/DESCRIPTION.md [hunk 6]
```diff
@@ -0,0 +1,6 @@
+---
+items:
+  - 1.jpeg
+---
+
+{DESCRIPTION}
```
Comment (leog) [inline packages/app-store/ouraring/DESCRIPTION.md:6]: Generated

File: packages/app-store/ouraring/api/add.ts [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+  const params = {
+    response_type: "code",
+    scope: scopes.join(" "),
+    client_id,
+    redirect_uri: WEBAPP_URL + "/api/integrations/ouraring/callback",
+    state,
+  };
+  const query = stringify(params);
+  const url = `https://cloud.ouraring.com/oauth/authorize?${query}`;
+  res.status(200).json({ url });
```
Comment (leog) [inline packages/app-store/ouraring/api/add.ts:30]: Standard add credentials flow for OAuth2

File: packages/app-store/ouraring/api/callback.ts [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+      userId: req.session?.user.id,
+      appId: "ouraring",
+    },
+  });
+  const state = decodeOAuthState(req);
+  return res.redirect(
+    getSafeRedirectUrl(state?.returnTo.replace("http:/", "http://")) ??
+      getInstalledAppPath({ variant: "other", slug: "ouraring" })
+  );
+}
```
Comment (leog) [inline packages/app-store/ouraring/api/callback.ts:72]: Standard callback flow for OAuth2

File: packages/app-store/ouraring/api/callback.ts [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+    data: {
+      type: "oura_ring_other_calendar",
+      key: responseBody,
+      userId: req.session?.user.id,
+      appId: "ouraring",
+    },
+  });
+  const state = decodeOAuthState(req);
+  return res.redirect(
+    getSafeRedirectUrl(state?.returnTo.replace("http:/", "http://")) ??
```
Comment (leog) [inline packages/app-store/ouraring/api/callback.ts:69]: For some reason returnTo has `http://` with a missing `/` so making sure it's there. Might need to do the same for `https`, will test in Cal-Staging (which is actually dev 😅 cc @keithwillcode)

File: packages/app-store/ouraring/lib/CalendarService.ts [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+
+  createEvent(): Promise<NewCalendarEventType | undefined> {
+    return Promise.resolve(undefined);
+  }
+  updateEvent(): Promise<NewCalendarEventType | NewCalendarEventType[]> {
+    return Promise.resolve([]);
+  }
+  deleteEvent(): Promise<unknown> {
+    return Promise.resolve();
+  }
```
Comment (leog) [inline packages/app-store/ouraring/lib/CalendarService.ts:71]: This calendar integration does not provide event creation interfaces, so we skip them as we usually do. Had to make `createEvent` kind of optional making it return `undefined` which was not allowed up until now, but it is computed as expected as it was an actual valid value.

File: packages/app-store/ouraring/lib/CalendarService.ts [hunk 11]
```diff
@@ -0,0 +1,180 @@
...
+                source: "Oura Ring",
+              },
+            ];
+          }
+          return [];
+        });
+        return Promise.resolve(result);
+      } else {
+        this.log.debug("No data returned");
+      }
```
Comment (leog) [inline packages/app-store/ouraring/lib/CalendarService.ts:104]: This is the esential code for what this app provides.

File: packages/app-store/ouraring/lib/CalendarService.ts [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+
+const refreshTokenResponseSchema = z.object({
+  access_token: z.string(),
+  expires_in: z
+    .number()
+    .transform((currentTimeOffsetInSeconds) => Math.round(+new Date() / 1000 + currentTimeOffsetInSeconds)),
+  refresh_token: z.string().optional(),
+});
+
+export default class OuraRingCalendarService implements Calendar {
```
Comment (leog) [inline packages/app-store/ouraring/lib/CalendarService.ts:50]: Usual OAuth2 flow to check existing `access_token` or generates a new one relying on `refresh_token`, nothing much to see here.

File: packages/core/CalendarManager.ts [hunk 13]
```diff
@@ -150,7 +150,7 @@ export const getCachedResults = async (
      * TODO: Migrate credential type or appId
      */
     const passedSelectedCalendars = selectedCalendars.filter((sc) => sc.integration === type);
-    if (!passedSelectedCalendars.length) return [];
+    if (!passedSelectedCalendars.length && type.indexOf("other_calendar") === -1) return [];
```
Comment (leog) [inline packages/core/CalendarManager.ts:153]: This change allows to run `getAvailability` for calendar app credentials that are not considered a valid calendar per se, such as this app.

File: packages/core/getCalendarsEvents.ts [hunk 14]
```diff
@@ -23,7 +23,7 @@ const getCalendarsEvents = async (
      * TODO: Migrate credential type or appId
      */
     const passedSelectedCalendars = selectedCalendars.filter((sc) => sc.integration === type);
-    if (!passedSelectedCalendars.length) return [];
+    if (!passedSelectedCalendars.length && type.indexOf("other_calendar") === -1) return [];
```
Comment (leog) [inline packages/core/getCalendarsEvents.ts:26]: This change allows to run `getAvailability` for calendar app credentials that are not considered a valid calendar per se, such as this app.

File: packages/lib/getSafeRedirectUrl.ts [hunk 15]
```diff
@@ -7,7 +7,10 @@ export const getSafeRedirectUrl = (url = "") => {
   }
 
   //It is important that this fn is given absolute URL because urls that don't start with HTTP can still deceive browser into redirecting to another domain
-  if (url.search(/^https?:\/\//) === -1) {
+  if (
+    (process.env.NODE_ENV === "production" && url.search(/^https?:/) === -1) ||
+    (process.env.NODE_ENV !== "production" && url.search(/^http?:/) === -1)
```
Comment (leog) [inline packages/lib/getSafeRedirectUrl.ts:12]: Checking also non-production environments relying on `http`

File: packages/prisma/seed-app-store.ts [hunk 16]
```diff
@@ -279,6 +279,12 @@ export default async function main() {
       base_url: (process.env.TANDEM_BASE_URL as string) || "https://tandem.chat",
     });
   }
+  if (process.env.OURA_RING_CLIENT_ID && process.env.OURA_RING_CLIENT_SECRET) {
+    await createApp("ouraring", "ouraring", ["other"], "oura_ring_other_calendar", {
+      client_id: process.env.OURA_RING_CLIENT_ID as string,
+      client_secret: process.env.OURA_RING_CLIENT_SECRET as string,
+    });
+  }
```
Comment (leog) [inline packages/prisma/seed-app-store.ts:287]: Adding this app to the list of apps in the DB based on the newly introduced environment variables defining client id and secret

File: packages/types/Calendar.d.ts [hunk 17]
```diff
@@ -215,7 +215,7 @@ export interface IntegrationCalendar extends Ensure<Partial<SelectedCalendar>, "
 }
 
 export interface Calendar {
-  createEvent(event: CalendarEvent): Promise<NewCalendarEventType>;
+  createEvent(event: CalendarEvent): Promise<NewCalendarEventType | undefined>;
```
Comment (leog) [inline packages/types/Calendar.d.ts:218]: This was already a valid value but it was not here so it couldn't be defined as `undefined` for non-traditional calendar apps such as this.

File: turbo.json [hunk 18]
```diff
@@ -239,6 +239,8 @@
     "NEXTAUTH_URL",
     "NODE_ENV",
     "ORGANIZATIONS_ENABLED",
+    "OURA_RING_CLIENT_ID",
+    "OURA_RING_CLIENT_SECRET",
```
Comment (leog) [inline turbo.json:243]: Newly introduced env vars for this app

**Comment by PeerRich** [comment]
awesome, lets review this on the weekends (since its a weekend project)

**Comment by PeerRich** [comment]
@leog would it be possible to merge this logic into the "Vital" app which already uses Oura Ring but also many other gadgets?

**Comment by leog** [comment]
> @leog would it be possible to merge this logic into the "Vital" app which already uses Oura Ring but also many other gadgets?

@PeerRich of course, I will do that when I have the time as a follow-up weekend project, closing this one.

**Pull request closed without merging**